### ISSUE TITLE
Dedicated VM stops execution after closing Colab tab

### ISSUE THREAD

**Issue by vpaul18** [googlecolab/colabtools#3205]
Hi! Colab seems to stop the execution approximately one hour after closing the tab. I do not expect this behavior since it is stated that "You can fully relax any runtime limits and idle timeouts by purchasing a dedicated VM at [GCP Marketplace]". 

Additionally, it is stated that  "Connecting to a custom GCE VM puts you in control of your machine lifecycle. You will still experience disconnections from your VM from interruptions to your connection, but Colab will not alter the VM state: your work and progress will be saved and available when you reconnect." 

Could you please let me know why I encounter this behavior & how to change it? Thank you!

**Comment by mco-gh** [comment]
Hi, Can you explain what you mean by "stop the execution"? If you have a running code cell, I wouldn't necessarily expect it to run indefinitely after closing a tab. But I would expect your server to preserve the state of a computation (imports, variables, functions, etc), as long as the serving VM isn't restarted for any reason.

If you can describe precisely the sequence of actions and expectations (and share an example notebook if possible), I can try to reproduce and understand what you're seeing.

**Comment by vpaul18** [comment]
Hello !  A very simple example of the expected outcome would be the following function printing the date as long as the underlying VM is running, regardless of whether or not the Colab tab is open or closed.

```
import time
import requests
import datetime

latest_date=datetime.datetime.now()

def print_date():
  while True:
    time.sleep(10)
    now = datetime.datetime.now()
    latest_date = now
    print(now)

print_date()
```

Instead, after one hour of printing the date- it stops. Additionally, I would expect the variable `latest_date` to persist the value from the run, but it does not. 

Thank you !

**Comment by mco-gh** [comment]
`latest_date` is a local variable in the `print_date` function. When you assign to it in that function, you're not changing the outer global value. So the global var is getting set only once and never changing.

To have the function update the global var, define it as global in the function definition (`global latest_date`).

**Comment by vpaul18** [comment]
@mco-gh  Ah alright, thanks for that!  What about the limited runtime, am I missing something there as well? Thank you!

**Comment by mco-gh** [comment]
I wouldn't equate a dropped or lost connection with a "limited runtime". I'd be inclined to set a variable to a known value, print it in a cell to verify, trigger a disconnect however you like (kill tab, reload page, let something time out, etc.), then reconnect the notebook and re-run the print to make sure the variable is still defined and still has the expected value. That tells you the runtime is persistent, even across lost connections.

**Comment by vpaul18** [comment]
@mco-gh  But is this an accurate way of testing whether the script stops running or not?

I have just performed the test you described and the variable persists, yet I'm not sure how representative is this outcome for my problem (have also tried the same test without the cloud engine instance & the same outcome was observed).

My problem is that I have a script that takes a long time to run and one hour after I close the tab it just stops executing. The way I know this is because I am always printing the progress ...

**Comment by mco-gh** [comment]
Understood, I'll try reproducing your experience and let you know what I see.

**Comment by mco-gh** [comment]
Ok, I ran your code overnight and it's still running now so something must be happening to cause your browser session to time out. Can you open Chrome devtools and see if you notice anything interesting in the JavaScript console or the network tab around the time your cell stops executing?

**Comment by mco-gh** [comment]
Sorry, I missed that you're not leaving it running, you're closing the tab. I'm now repeating the repro experiment with killing the running tab to see if I get the same result.

**Comment by mco-gh** [comment]
I ran this test last night, killed the tab right after starting the cell, and reconnected this morning (w/ a Marketplace VM), and this morning I was able to reconnect and it picked up right where I left off so this is working for me.

There were some connection issues fixed recently with the Marketplace VM. Can you try to reproduce this yourself? If you see it again, can you ssh into your VM and run `docker ps -a`? I'd like to see if one of our containers is dying when you experience the timeout.

**Comment by vpaul18** [comment]
Hi Marc, thanks a lot for looking into it! I have been creating new notebooks, doing more trials & it seems like my problem arose from the fact that one has to connect each notebook to the VM.  

I think the confusion was given by the fact that the tab containing the VM details was always pre-filled with the VM details and hovering over the environment stated something like 'connected to Compute Engine VM'. 

In short: I was not aware one has to connect to the custom environment within each notebook. Everything seems to work as expected now. Thanks a lot for your time!

**Comment by mco-gh** [comment]
That's great, glad it's sorted! Thanks for letting me know and I'll close this issue.
